File: app/src/main/java/com/cyl/musiclake/ui/localmusic/presenter/MyMusicPresenter.java
package com.cyl.musiclake.ui.localmusic.presenter;

import android.content.Context;

import com.cyl.musiclake.data.model.Music;
import com.cyl.musiclake.data.model.Playlist;
import com.cyl.musiclake.data.source.AppRepository;
import com.cyl.musiclake.data.source.PlaylistLoader;
import com.cyl.musiclake.ui.localmusic.contract.MyMusicContract;

import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by yonglong on 2018/1/6.
 */

public class MyMusicPresenter implements MyMusicContract.Presenter {
    private MyMusicContract.View mView;
    private Context mContext;
    private List<Playlist> playlists;
    private List<Music> musicList;

    public MyMusicPresenter(Context mContext) {
        this.mContext = mContext;
    }

    @Override
    public void attachView(MyMusicContract.View view) {
        mView = view;
    }

    @Override
    public void subscribe() {
    }

    @Override
    public void unsubscribe() {

    }

    @Override
    public void loadSongs() {
        AppRepository.getAllSongsRepository(mContext)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Music>>() {

                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(List<Music> musicList) {
                        mView.showSongs(musicList);
                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                        mView.hideLoading();
                    }

                    @Override
                    public void onComplete() {
                        mView.hideLoading();
                    }
                });

        AppRepository.getFavoriteSong(mContext)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Music>>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(List<Music> musicList) {
                        mView.showLoveList(musicList);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });

        AppRepository.getPlaylistSongsRepository(mContext, "1")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Music>>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(List<Music> musicList) {
                        mView.showHistory(musicList);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void loadPlaylist() {
        playlists = PlaylistLoader.getPlaylist(mContext);
        mView.showPlaylist(playlists);
        if (playlists.size() == 0) {
            mView.showEmptyView();
        }
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/my/MyMusicPresenter.kt
package com.cyl.musiclake.ui.music.my

import com.cyl.musiclake.api.music.netease.NeteaseApiServiceImpl
import com.cyl.musiclake.api.net.ApiManager
import com.cyl.musiclake.api.net.RequestCallBack
import com.cyl.musiclake.bean.Playlist
import com.cyl.musiclake.data.PlayHistoryLoader
import com.cyl.musiclake.data.PlaylistLoader
import com.cyl.musiclake.data.SongLoader
import com.cyl.musiclake.data.VideoLoader
import com.cyl.musiclake.ui.music.edit.PlaylistManagerUtils
import com.cyl.musiclake.ui.base.BasePresenter
import com.cyl.musiclake.ui.download.DownloadLoader
import com.cyl.musiclake.ui.my.user.UserStatus
import com.cyl.musiclake.utils.LogUtil
import com.cyl.musiclake.utils.SPUtils
import com.cyl.musiclake.utils.ToastUtils
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import javax.inject.Inject

/**
 * Created by yonglong on 2018/1/6.
 */

class MyMusicPresenter @Inject
constructor() : BasePresenter<MyMusicContract.View>(), MyMusicContract.Presenter {
    /**
     * 更新播放历史
     */
    fun updateHistory() {
        doAsync {
            val data = PlayHistoryLoader.getPlayHistory()
            uiThread {
                mView?.showHistory(data)
            }
        }
    }

    /**
     * 更新播放历史
     */
    private fun updateLocal() {
        doAsync {
            val data = SongLoader.getLocalMusic(mView.context)
            uiThread {
                mView?.showSongs(data)
            }
        }
    }

    /**
     * 更新本地歌单
     */
    fun updateFavorite() {
        doAsync {
            val data = SongLoader.getFavoriteSong()
            uiThread {
                mView?.showLoveList(data)
            }
        }
    }


    /**
     * 更新下载歌曲
     */
    fun updateDownload() {
        doAsync {
            val data = DownloadLoader.getDownloadList()
            uiThread {
                mView?.showDownloadList(data)
            }
        }
    }

    /**
     * 更新下载歌曲
     */
    fun updateLocalVideo() {
        doAsync {
            val musicList = VideoLoader.getAllLocalVideos(mView.context)
            uiThread {
                mView?.showVideoList(musicList)
            }
        }
    }


    override fun loadSongs() {
        updateLocal()
        updateHistory()
        updateFavorite()
        updateLocalVideo()
        updateDownload()
    }

    fun loadMusicLakeNotice() {
        PlaylistManagerUtils.getMusicNoticeInfo(
                success = {
                    mView?.showNoticeInfo(it)
                }, fail = {
        })
    }

    /**
     * 获取在线歌单
     */
    override fun loadPlaylist(playlist: Playlist?) {
        if (UserStatus.getLoginStatus() && UserStatus.getTokenStatus()) {
            val mIsLogin = UserStatus.getLoginStatus()
            if (mIsLogin) {
                PlaylistManagerUtils.getOnlinePlaylist(success = {
                    mView?.showPlaylist(it)
                }, fail = {
                    ToastUtils.show(it)
                    if (PlaylistManagerUtils.playlists.size == 0) {
                        mView?.showError(it, true)
                    }
                })
            } else {
                PlaylistManagerUtils.playlists.clear()
                mView?.showPlaylist(PlaylistManagerUtils.playlists)
            }
        } else {
            mView?.showPlaylist(mutableListOf())
        }
        loadWyUserPlaylist()
    }

    /**
     * 获取本地歌单
     */
    fun loadLocalPlaylist() {
        doAsync {
            val playlist = PlaylistLoader.getAllPlaylist()
            playlist.forEach {
                it.pid?.let { it1 ->
                    val list = PlaylistLoader.getMusicForPlaylist(it1)
                    it.total = list.size.toLong()
                }
            }
            uiThread {
                mView?.showLocalPlaylist(playlist)
            }
        }
    }


    /**
     * 加载网易云歌单
     */
    fun loadWyUserPlaylist() {
        val uid = SPUtils.getAnyByKey(SPUtils.SP_KEY_NETEASE_UID, "")
        LogUtil.d("MyMusic", "uid = $uid")
        if (uid.isEmpty()) return
        val observable = NeteaseApiServiceImpl.getUserPlaylist(uid = uid)
        ApiManager.request(observable, object : RequestCallBack<MutableList<Playlist>> {
            override fun success(result: MutableList<Playlist>) {
                mView?.showWyPlaylist(result)
            }

            override fun error(msg: String) {
                mView?.showWyPlaylist(mutableListOf())
            }
        })
    }
}
